mne/beamformer/_ap.py

```diff
@@ -0,0 +1,712 @@
+"""compute an Alternating Projection (AP)."""
+
+# Authors: Yuval Realpe <yuval.realpe@gmail.com>
+#
+# License: BSD-3-Clause
+
+from copy import copy
+
+import numpy as np
+from numpy.linalg import pinv, multi_dot, lstsq
+
+from ..utils import _check_info_inv, verbose, fill_doc, logger
+from ._compute_beamformer import _prepare_beamformer_input
+from ..io.pick import pick_channels_forward, pick_info
+from ..forward.forward import convert_forward_solution, is_fixed_orient
+from ..inverse_sparse.mxne_inverse import _make_dipoles_sparse
+from ..minimum_norm.inverse import _log_exp_var
+
+
+def _produce_data_cov(data_arr, attr_dict):
+    """Calculate data Covariance."""
+    nsources = attr_dict['nsources']
+    data_tr = data_arr.transpose()
+    data_norm = np.matmul(data_arr, data_tr)
+    data_cov = (data_norm + data_norm.trace()
+                * np.eye(data_arr.shape[0]) * 0.001)  # Array Covariance Matrix
+    logger.info(' alternating projection ; nsources = {}:'.format(nsources))
+
+    return data_cov
+
+
+def _active_subspace(Mat):
+    """Calculate Multi_dot with pseudo-inverse."""
+    return multi_dot([Mat, pinv(Mat.T @ Mat), Mat.T])
+
+
+def _fixed_phase1a(attr_dict, data_cov, gain):
+    """Calculate phase 1a of fixed oriented AP.
+
+    Initialization: search the 1st source location over the entire
+    dipoles topographies space.
+
+    Parameters
+    ----------
+    attr_dict : dict
+        See: _calculate_fixed_alternating_projections.
+    data_cov : array
+        Data Covariance.
+    gain : array, shape (nchannels, ndipoles)
+        Gain matrix.
+
+    Returns
+    -------
+    s_ap : list of int
+        List of dipole indices.
+
+    """
+    s_ap = []
+    ap_val1 = np.zeros(attr_dict['ndipoles'])
+    for dip in range(attr_dict['ndipoles']):
+        l_p = gain[:, [dip]]
```

Comment:
```suggestion
        l_p = gain[:, dip:dip + 1]
```
to avoid a memory copy of the column of gain.

Now this for loop should be easily avoidable.

Something like:

`ap_val1 = np.sum(gain * (data_cov @ gain), axis=0) / np.linalg.norm(gain, axis=0)**2`

Reply:
I would also recommend to avoid passing the dict attr_dict and just passe explicitly the parameters needed. It makes the code less opaque. Here for example it would mean replacing attr_dict by ndipoles so not adding a ton of paramers

Reply:
avoided all relevant loops, specifically the ones found on _fixed_phaseX functions.
retired the attr_dict completely .
changed gain[:, [dip]] --> gain[:, dip:dip+1]